stan/math/rev/fun/singular_values.hpp

```diff
@@ -0,0 +1,77 @@
+#ifndef MATH_SINGULAR_VALUES_H
+#define MATH_SINGULAR_VALUES_H
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/functor/reverse_pass_callback.hpp>
+#include <stan/math/rev/core/arena_matrix.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/rev/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <Eigen/SVD>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the singular values of the specified symmetric matrix.
+ *
+ * For MxN input matrix, M must be <= N
+ *
+ * @tparam EigMat type of input matrix
+ * @param m MxN input matrix, M <= N
+ * @return Singular values of matrix
+ */
+template <typename EigMat,
+	  require_eigen_vt<EigMat>* = nullptr,
+	  require_not_vt_var<EigMat>* = nullptr>
+inline auto singular_values(const EigMat& m) {
+  using ret_type = promote_scalar_t<var, Eigen::VectorXd>;
+  check_nonzero_size("singular_values", "m", m);
+
+  const int M = m.rows();
+  auto arena_m = to_arena(m);
+  
+  Eigen::MatrixXd m_val = value_of(arena_m);
+  Eigen::JacobiSVD<EigMat> svd(m_val, Eigen::ComputeFullU | Eigen::ComputeFullV);
+
+  Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic> D = Eigen::MatrixXd::Zero(M,M);
+  D.diagonal() = svd.singularValues();
+  arena_t<ret_type> sigular_values = svd.singularValues();
+
+  auto U = svd.matrixU();
+  auto V = svd.matrixU();
+
+  // equation (4) from "Differentiable Programming Tensor Networks"
+  Eigen::MatrixXd Fp(M,M);
+  Eigen::MatrixXd Fm(M,M);
+
+  for(int i = 0; i < M; i++) {
+    for(int j = 0; i < M; j++) {
+      if(j == i) {
+	Fp[i, j] = 0.0;
+	Fm[i, j] = 0.0;
+      } else {
+	Fp[i, j] = 1.0 / (D[j,j] - D[i,i]) + 1.0 / (D[i,i] + D[j,j]);
+	Fm[i, j] = 1.0 / (D[j,j] - D[i,i]) - 1.0 / (D[i,i] + D[j,j]);
+      }
+    }
+  }
```

Comment:
`Fp` and `Fm` are not necessary for propagating the adjoints of the singular values. Since the output here are the singular values, that's all we need to worry about.

Reply:
Fp and Fm are unused in this calculation